app/src/main/kotlin/com/simplemobiletools/musicplayer/helpers/MultiPlayer.kt

```diff
@@ -0,0 +1,288 @@
+            AudioManager.AUDIOFOCUS_GAIN -> {
+                if (!isPlaying() && isPausedByTransientLossOfFocus) {
+                    start()
+                    callbacks.onPlayStateChanged()
+                    isPausedByTransientLossOfFocus = false
+                }
+                setVolume(Volume.NORMAL)
+            }
+            AudioManager.AUDIOFOCUS_LOSS -> {
+                pause()
+                callbacks.onPlayStateChanged()
+            }
+            AudioManager.AUDIOFOCUS_LOSS_TRANSIENT -> {
+                val wasPlaying = isPlaying()
+                pause()
+                callbacks.onPlayStateChanged()
+                isPausedByTransientLossOfFocus = wasPlaying
+            }
+            AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK -> {
+                setVolume(Volume.DUCK)
+            }
+        }
+    }
+
+    private val audioFocusRequest: AudioFocusRequestCompat =
+        AudioFocusRequestCompat.Builder(AudioManagerCompat.AUDIOFOCUS_GAIN).setOnAudioFocusChangeListener(audioFocusListener).setAudioAttributes(
+            AudioAttributesCompat.Builder().setContentType(AudioAttributesCompat.CONTENT_TYPE_MUSIC).build()
+        ).build()
+
+    init {
+        mCurrentMediaPlayer.setWakeMode(app, PowerManager.PARTIAL_WAKE_LOCK)
+    }
+
+    fun getAudioSessionId(): Int = mCurrentMediaPlayer.audioSessionId
+
+    fun isPlaying(): Boolean = isInitialized && mCurrentMediaPlayer.isPlaying
+
+    fun setDataSource(trackUri: Uri) {
+        isInitialized = false
+        setDataSourceImpl(mCurrentMediaPlayer, trackUri) { result ->
+            if (result.isFailure) {
+                throw result.exceptionOrNull()!!
+            }
+            isInitialized = true
+            setNextDataSource(null)
+            callbacks.onPrepared()
+        }
+    }
+
+    fun setNextDataSource(trackUri: Uri?, onPrepared: (() -> Unit)? = null) {
+        try {
+            mCurrentMediaPlayer.setNextMediaPlayer(null)
+        } catch (e: IllegalArgumentException) {
+            // Next media player is current one, continuing
+        } catch (e: IllegalStateException) {
+            return
+        }
+        if (mNextMediaPlayer != null) {
+            mNextMediaPlayer?.release()
+            mNextMediaPlayer = null
+        }
+        if (trackUri == null) {
+            return
+        }
+        if (app.config.gapLessPlayback) {
+            mNextMediaPlayer = MediaPlayer()
+            mNextMediaPlayer?.setWakeMode(app, PowerManager.PARTIAL_WAKE_LOCK)
+            mNextMediaPlayer?.audioSessionId = mCurrentMediaPlayer.audioSessionId
+            setDataSourceImpl(mNextMediaPlayer!!, trackUri) { result ->
+                if (result.isSuccess) {
+                    try {
+                        mCurrentMediaPlayer.setNextMediaPlayer(mNextMediaPlayer)
+                        onPrepared?.invoke()
+                    } catch (e: IllegalArgumentException) {
+                        if (mNextMediaPlayer != null) {
```

Comment:
null checks arent really needed here, could be simplified

Reply:
simplified.